core/vm/instructions.go

```diff
@@ -943,6 +943,22 @@ func opSelfdestruct6780(pc *uint64, interpreter *EVMInterpreter, scope *ScopeCon
 		tracer.CaptureEnter(SELFDESTRUCT, scope.Contract.Address(), beneficiary.Bytes20(), []byte{}, 0, balance)
 		tracer.CaptureExit([]byte{}, 0, nil)
 	}
+	if interpreter.evm.chainRules.IsPrague {
+		contractAddr := scope.Contract.Address()
+		beneficiaryAddr := beneficiary.Bytes20()
+		// If the beneficiary isn't the contract, we need to touch the beneficiary's balance.
+		if contractAddr != beneficiaryAddr {
+			statelessGas := interpreter.evm.Accesses.TouchAddressOnReadAndComputeGas(beneficiaryAddr[:], uint256.Int{}, trieUtils.BalanceLeafKey)
+			if !scope.Contract.UseGas(statelessGas) {
+				scope.Contract.Gas = 0
+				return nil, ErrOutOfGas
+			}
+		}
+		// If the beneficiary is the contract itself, we *don't touch* since there're two possibilities:
+		// 1. The contract was created in the same transaction: the balance is already touched.
+		// 2. The contract wasn't created in the same transaction: there's no net change in balance,
```

Comment:
there's no change in the balance, but it is read as seen line 940, so it should go in there I think.

Reply:
We could avoid this `AddBalance + SubBalance` as a gas optimization -- let's discuss that in another separate talk. I'll change to touch it in that case.